flink-libraries/flink-gelly/src/main/java/org/apache/flink/graph/BipartiteGraph.java

```diff
@@ -0,0 +1,364 @@
+	 */
+	public Graph<KB, VVB, Tuple2<EV, EV>> projectionBottomSimple() {
+
+		DataSet<Edge<KB, Tuple2<EV, EV>>> newEdges =  edges.join(edges)
+			.where(0)
+			.equalTo(0)
+			.filter(new FilterFunction<Tuple2<BipartiteEdge<KT, KB, EV>, BipartiteEdge<KT, KB, EV>>>() {
+				@Override
+				public boolean filter(Tuple2<BipartiteEdge<KT, KB, EV>, BipartiteEdge<KT, KB, EV>> value) throws Exception {
+					BipartiteEdge<KT, KB, EV> edge1 = value.f0;
+					BipartiteEdge<KT, KB, EV> edge2 = value.f1;
+					return !edge1.getBottomId().equals(edge2.getBottomId());
+				}
+			})
+			.map(new MapFunction<Tuple2<BipartiteEdge<KT, KB, EV>, BipartiteEdge<KT, KB, EV>>, Edge<KB, Tuple2<EV, EV>>>() {
+				@Override
+				public Edge<KB, Tuple2<EV, EV>> map(Tuple2<BipartiteEdge<KT, KB, EV>, BipartiteEdge<KT, KB, EV>> value) throws Exception {
+					return new Edge<>(
+						value.f0.getBottomId(),
+						value.f1.getBottomId(),
+						new Tuple2<>(
+							value.f0.getValue(),
+							value.f1.getValue()));
+				}
+			});
+
+		return Graph.fromDataSet(bottomVertices, newEdges, context);
+	}
+
+	/**
+	 * Convert a bipartite into a graph that contains only bottom vertices. An edge between two vertices in the new
+	 * graph will exist only if the original bipartite graph contains a top vertex they both connected to.
+	 *
+	 * @return bottom projection of the bipartite graph where every edge contains a data about projected connection
+	 * between vertices in the original graph
+	 */
+	public Graph<KB, VVB, Projection<KT, VVT, EV, VVB>> projectionBottomFull() {
+
+		// Join edges with bottom vertices to preserve edges values
+		DataSet<Tuple2<BipartiteEdge<KT, KB, EV>, Vertex<KB, VVB>>> ds = edges.join(bottomVertices)
+			.where(1)
+			.equalTo(0);
+
+		DataSet<Edge<KB, Projection<KT, VVT, EV, VVB>>> newEdges = ds.join(ds)
+			.where(new KeySelector<Tuple2<BipartiteEdge<KT,KB,EV>,Vertex<KB,VVB>>, KT>() {
+				@Override
+				public KT getKey(Tuple2<BipartiteEdge<KT, KB, EV>, Vertex<KB, VVB>> value) throws Exception {
+					return value.f0.getTopId();
+				}
+			})
+			.equalTo(new KeySelector<Tuple2<BipartiteEdge<KT,KB,EV>,Vertex<KB,VVB>>, KT>() {
+				@Override
+				public KT getKey(Tuple2<BipartiteEdge<KT, KB, EV>, Vertex<KB, VVB>> value) throws Exception {
+					return value.f0.getTopId();
+				}
+			})
+			// Filter all pairs with the same bottom id
+			.filter(new FilterFunction<Tuple2<
+					Tuple2<BipartiteEdge<KT, KB, EV>, Vertex<KB, VVB>>,
+					Tuple2<BipartiteEdge<KT, KB, EV>, Vertex<KB, VVB>>>>() {
+				@Override
+				public boolean filter(Tuple2<
+						Tuple2<BipartiteEdge<KT, KB, EV>, Vertex<KB, VVB>>,
+						Tuple2<BipartiteEdge<KT, KB, EV>, Vertex<KB, VVB>>> value) throws Exception {
+					BipartiteEdge<KT, KB, EV> edge1 = value.f0.f0;
+					BipartiteEdge<KT, KB, EV> edge2 = value.f1.f0;
+					return !edge1.getBottomId().equals(edge2.getBottomId());
+				}
+			})
+			// Join with top vertices to preserve top vertices values
+			.join(topVertices)
```

Comment:
Is it not more efficient to do the Cartesian join last if we assume that the |vertices| << |bipartite edges| << |simple edges|? The code can be reused between the full projection functions: first join the bipartite edge with top vertex, then join the result with the bottom vertex (perhaps using the Projection class below with NullValue where appropriate).